### ISSUE TITLE
Request: Disallow iteration in render

### ISSUE THREAD

**Issue by TheSavior** [yannickcr/eslint-plugin-react#651]
What are your thoughts on a rule like this:

Invalid

```
React.createClass({
  propTypes: {
    array: React.PropTypes.array
  },

  render: function() {
    return (
      <div>
      {
        this.props.array.map((item) => <MyComponent {...item})
      }
      </div>
    );
  }
});
```

Valid

```
React.createClass({
  propTypes: {
    array: React.PropTypes.array
  },

  render: function() {
    return (
      <div>
      {this.renderItems()}
      </div>
    );
  },

  renderItems: function() {
    return this.props.array.map((item) => <MyComponent {...item});
  }
});
```

**Comment by ljharb** [comment]
Why is this preferable? Personally I much prefer the former, because then I can make the component an SFC - since it has no state.

**Comment by TheSavior** [comment]
Our team likes and enforces this rule because we find it easier and quicker to read the render function to see what a component is doing by seeing the well named function calls instead of inline maps.

Real code sample from our repo:

```
React.createClass({
  // ...
  render() {
    return (
      <div>
        <table>
          <tbody>
            {this.renderCashRows()}
            {this.renderTransferRows()}
            {this.renderUnsureRows()}
            {this.renderUnsupportedRows()}
          </tbody>
        </table>
        {this.renderTaxConsequences()}
      </div>
    );
  }
  // ...
});
```

We find it way more readable to see that than to see the inline maps those functions are doing. Because of this, we socially enforce it everywhere, but want to be able to lint for it as well.

**Comment by TheSavior** [comment]
@ljharb, since you are a 👎  on the PR, can you provide some more insight into your thoughts on the code sample in my above comment? I'd like to better understand how you would approach that type of render with multiple iterations. I'm sure I'll learn something.

What if the calls to map weren't just passing an item to a prop, what if the map statements were many lines doing some work on each item and then passing the results as props to a child item? What if those functions had `if` statements in them to conditionally return out instead of doing the map? Do you still believe that work should be done directly in the render function?

**Comment by ljharb** [comment]
Here's how I'd write it:
```js
function Component({ array }) {
   return (
     <div>
       {array.map(item => (
         <MyComponent {...item} />
       ))}
     </div>
  );
}
Component.propTypes = {
  array: React.PropTypes.array,
};
```

If a component is complex, partially hiding that complexity by burying it in instance methods is a poor solution imo. When "everything inline" is hard to read, that's a strong indicator that your component hierarchy needs to be altered - make lots of small components, rather than big complex ones.

**Comment by TheSavior** [comment]
I agree that small components are much better. Easier to understand, easier to test, easier to reuse. 

However, it isn't always possible, tables are one example since a component can't return multiple `tr`s. 

The point here isn't that because tables cause a problem we should give up on small components, just that there are many different sides to this story and tradeoffs to any given approach.

**Comment by ljharb** [comment]
Closing, per https://github.com/yannickcr/eslint-plugin-react/pull/652#issuecomment-229261462.
